src/models/Integration.php

```diff
@@ -0,0 +1,379 @@
+
+        $registration_facebook =
+          await Configuration::gen('registration_facebook');
+        if ($oauth_token_exists === true) {
+          $url = await self::genLoginURL("facebook_oauth", $email);
+        } else if ($registration_facebook->getValue() === '1') {
+          $team_id = await self::genRegisterTeam($email, $id);
+          if (is_int($team_id) === true) {
+            $set_integrations = await self::genSetTeamIntegrations(
+              $team_id,
+              'facebook_oauth',
+              $email,
+              $id,
+            );
+            if ($set_integrations === true) {
+              $url = await self::genLoginURL('facebook_oauth', $email);
+            }
+          }
+        }
+      }
+    } else if ($error !== false) {
+      $url = '/index.php?page=login';
+    }
+
+    return $url;
+  }
+
+  public static async function genGoogleLogin(): Awaitable<string> {
+    list($client, $url) = await self::genGoogleAuthURL("login");
+
+    $code = idx(Utils::getGET(), 'code');
+    $error = idx(Utils::getGET(), 'error');
+
+    if (!is_string($code)) {
+      $code = false;
+    }
+
+    if (!is_string($error)) {
+      $error = false;
+    }
+
+    if ($code !== false) {
+      $url = '/index.php?page=login';
+      $client->authenticate($code);
+      $access_token = $client->getAccessToken();
+      $oauth_client = new Google_Service_Oauth2($client);
+      $profile = $oauth_client->userinfo->get();
+      $email = $profile->email;
+      $id = $profile->id;
+
+      $oauth_token_exists =
+        await Team::genAuthTokenExists('google_oauth', strval($email));
+
+      $registration_google = await Configuration::gen('registration_google');
+      if ($oauth_token_exists === true) {
+        $url = await self::genLoginURL('google_oauth', $email);
+      } else if ($registration_google->getValue() === '1') {
+        $team_id = await self::genRegisterTeam($email, $id);
+        if (is_int($team_id) === true) {
+          $set_integrations = await self::genSetTeamIntegrations(
+            $team_id,
+            'google_oauth',
+            $email,
+            $id,
+          );
+          if ($set_integrations === true) {
+            $url = await self::genLoginURL('google_oauth', $email);
+          }
+        }
+      }
+    } else if ($error !== false) {
+      $url = '/index.php?page=login';
+    }
+
+    return $url;
+  }
+
+  public static async function genLoginURL(
+    string $type,
+    string $token,
+  ): Awaitable<string> {
+    $team = await Team::genTeamFromOAuthToken($type, $token);
+
+    SessionUtils::sessionRefresh();
+    if (!SessionUtils::sessionActive()) {
+      SessionUtils::sessionSet('team_id', strval($team->getId()));
+      SessionUtils::sessionSet('name', $team->getName());
+      SessionUtils::sessionSet(
+        'csrf_token',
+        (string) gmp_strval(
+          gmp_init(bin2hex(openssl_random_pseudo_bytes(16)), 16),
+          62,
+        ),
```

Comment:
- use random_bytes() instead of openssl_random_pseudo_bytes()
 - similar: why base62 with gmp instead of base64 or hex?